group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -761,6 +776,58 @@ public ClassicGroup classicGroup(
         }
     }
 
+    /**
+     * Validates the online upgrade if the Classic Group receives a ConsumerGroupHeartbeat request.
+     *
+     * @param classicGroup A ClassicGroup.
+     * @return the boolean indicating whether it's valid to online upgrade the classic group.
+     */
+    private boolean validateOnlineUpgrade(ClassicGroup classicGroup) {
+        if (!consumerGroupMigrationPolicy.isUpgradeEnabled()) {
+            log.debug("Online upgrade is invalid because the consumer group {} migration config is {} so online upgrade is not enabled.",
+                classicGroup.groupId(), consumerGroupMigrationPolicy);
+            return false;
+        } else if (classicGroup.isInState(DEAD)) {
```

Comment:
Could this really happen? I would have thought that it would be automatically converted as Dead or equivalent to Empty.

Reply:
Yeah it makes sense. Actually I think we don't set classicGroup to dead at all. We can delete the else if